### ISSUE TITLE
stats: document server-side stats event ordering and add tests

### ISSUE THREAD

**Issue by aranjans** [grpc/grpc-go#7824]
## Description
Currently, the ordering of server-side stats events is not explicitly documented in the `stats.Handler` interface, nor verified through tests. This can lead to confusion for users implementing stats handlers, as seen in [link-to-user-question-if-available].

The current server-side stats event ordering is:
1. InHeader (when processing incoming headers)
2. Begin (when starting RPC processing)
3. [message events during RPC]
4. End (when RPC completes)

## Proposed Changes
1. Update the `stats.Handler` interface documentation to explicitly state the ordering of server-side stats events
2. Add tests to verify this behavior remains consistent

The current behavior is correct and working as intended, but lacks documentation and test coverage to ensure it remains consistent. This will help users better understand and rely on the stats event ordering when implementing their stats handlers.

**Comment by dfawley** [comment]
> The expected server-side stats event ordering is:

This is the _current_ ordering.

We should also discuss options like:

- Removing "Begin", since the start of every RPC on either side should include the initial headers
- Omitting "InHeader"/"OutHeader" at RPC start, since the start of every RPC on either side should already include the headers
- Reordering the events, since logically "Begin" sounds like it should happen first

I'm not wild about the current behavior.  But changing things here may be difficult since the API has been around for a long time.

**Comment by arjan-bal** [comment]
@RyanBlaney thank you for taking this up. The proposed changes sound good at a high level. I've started reviewing #7885.

**Comment by dfawley** [comment]
Note that the comments you've added in the above snippets only apply to the _server_.  On the client, I believe we expect this ordering: `Begin->OutHeader->{InHeader (at most once)/OutPayload/InPayload (only after InHeader)}->InTrailer`.

**Comment by RyanBlaney** [comment]
Yes thank you! That clears up a lot. I encountered that sequence for the client during testing and was confused. I clarify in the documentation to address this. I'll most likely add more test cases to cover this.

**Comment by RyanBlaney** [comment]
Here's how the documentation looks in Neovim:
![image](https://github.com/user-attachments/assets/4ccb7d71-0a40-4753-b070-ccd7168b731c)

The single space indent allows for the nice format, let me know if I should remove it for consistency with other functions.

```go
// InHeader contain stats when the header is received.
//  First event in the server side event sequence.
//  Follows last OutPayload for server side events.
// 
// Server Stats Example Event Ordering:
//  *InHeader* -> Begin -> InPayload(s) -> OutHeader -> OutPayload(s) -> OutTrailer -> End 
// 
// Client Stats Example Event Ordering:
//  Begin -> OutHeader -> OutPayload(s) -> *InHeader* -> InTrailer -> InPayload(s) -> End 
```

Here is Begin:
![image](https://github.com/user-attachments/assets/92276ad6-f326-4651-b7e7-7cbe62f8e740)

```go
// Begin contains stats when an RPC attempt begins.
//  First event in the client-side event sequence.
//  Follows InHeader for server-side events.
// 
// Server Stats Example Event Ordering:
//  InHeader -> *Begin* -> InPayload(s) -> OutHeader -> OutPayload(s) -> OutTrailer -> End
// 
// Client Stats Example Event Ordering:
//  *Begin* -> OutHeader -> OutPayload(s) -> InHeader -> InTrailer -> InPayload(s) -> End
//
// FailFast is only valid if this Begin is from client side.
```

**Comment by arjan-bal** [comment]
PR: https://github.com/grpc/grpc-go/pull/7885
